src-java/blue-green/src/main/java/org/openkilda/bluegreen/ZkWatchDog.java

```diff
@@ -0,0 +1,127 @@
+/* Copyright 2020 Telstra Open Source
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ */
+
+
+package org.openkilda.bluegreen;
+
+import com.google.common.annotations.VisibleForTesting;
+import lombok.Builder;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.zookeeper.AsyncCallback.DataCallback;
+import org.apache.zookeeper.KeeperException;
+import org.apache.zookeeper.WatchedEvent;
+import org.apache.zookeeper.data.Stat;
+
+import java.io.IOException;
+import java.util.HashSet;
+import java.util.Set;
+
+@Slf4j
+public class ZkWatchDog extends ZkClient implements WatchDog, DataCallback {
+
+    private static final String SIGNAL = "signal";
+    @VisibleForTesting
+    protected String signalPath;
+    private Signal signal;
+
+    @VisibleForTesting
+    Set<LifeCycleObserver> observers = new HashSet<>();
+
+    @Builder
+    public ZkWatchDog(String id, String serviceName, String connectionString,
+                      int sessionTimeout, Signal signal) throws IOException {
+        super(id, serviceName, connectionString, sessionTimeout);
+
+        signalPath = getPaths(serviceName, id, SIGNAL);
+        if (signal == null) {
+            signal = Signal.NONE;
+        }
+        this.signal = signal;
+        initWatch();
```

Comment:
Rearrange `initWatch()` calls to be a part of successful ZK connection processing. Now... it called at least in each(?) constructor of ZkClient and in some way during reconnects.

PS Maybe it would be renamed in something like `postConnect`.